kubernetes-membership-scheme/src/main/java/org/wso2/carbon/membership/scheme/kubernetes/KubernetesMembershipScheme.java

```diff
@@ -68,42 +70,64 @@ public KubernetesMembershipScheme(Map<String, Parameter> parameters, String prim
         this.carbonCluster = hazelcastCarbonCluster;
     }
 
+    /**
+     * Retrieves the set of IP addresses of the current PODs in the K8S cluster, using K8S API server or the DNS.
+     * The returned set may contain the IP address of the pod running this JVM.
+     *
+     * The IP address resolver will use DNS lookup if "USE_DNS" environment property is set. Otherwise it uses K8S
+     * API server, to retrive the IP addresses.
+     *
+     * @return IP addresses of the current pods. Returns null upon any error querying K8S.
+     */
+    private Set<String> getK8sPodIpAddresses() throws KubernetesMembershipSchemeException {
+        Set<String> containerIps = podIpResolver.resolveAddresses();
+        if (containerIps != null) {
+            return containerIps;
+        } else {
+            return Collections.emptySet();
+        }
+    }
+
+    /**
+     * Initiates the Pod IP resolver.
+     * Uses the DNS based pod IP resolver or the API based pod IP resolver.
+     */
+    private void initPodIpResolver() throws KubernetesMembershipSchemeException {
+        String useDns = System.getenv(Constants.USE_DNS);
+        if (StringUtils.isEmpty(useDns)) {
+            useDns = getParameterValue(Constants.USE_DNS, "true");
+        }
+        if (Boolean.parseBoolean(useDns)) {
+            log.debug("Using DNS based pod ip resolving method");
+            podIpResolver = new DNSBasedPodIpResolver(parameters);
+        } else {
+            log.debug("Using API based pod ip resolving method");
+            podIpResolver = new ApiBasedPodIpResolver(parameters);
+        }
+    }
+
     @Override public void init() throws ClusteringFault {
         try {
             log.info("Initializing kubernetes membership scheme...");
-
             nwConfig.getJoin().getMulticastConfig().setEnabled(false);
             nwConfig.getJoin().getAwsConfig().setEnabled(false);
             TcpIpConfig tcpIpConfig = nwConfig.getJoin().getTcpIpConfig();
             tcpIpConfig.setEnabled(true);
-
-            String useDns = System.getenv(Constants.USE_DNS);
-            if (StringUtils.isEmpty(useDns)) {
-                useDns = getParameterValue(Constants.USE_DNS, "true");
-            }
-
-            AddressResolver podIpResolver;
-            Set<String> containerIPs;
-            if (Boolean.parseBoolean(useDns)) {
-                log.debug("Using DNS based pod ip resolving method");
-                podIpResolver = new DNSBasedPodIpResolver(parameters);
-            } else {
-                log.debug("Using API based pod ip resolving method");
-                podIpResolver = new ApiBasedPodIpResolver(parameters);
-            }
-            containerIPs = podIpResolver.resolveAddresses();
+            initPodIpResolver();
+            Set <String> containerIPs = getK8sPodIpAddresses();
             // if no IPs are found, can't initialize clustering
             if (containerIPs.isEmpty()) {
```

Comment:
Can this happen? If this member (local member) is running, it means there should be at least one pod. Anyway there is no harm in having this condition.

Reply:
Yes. Since there is already a pod running the IP list wouldn't be empty.